[로또 2단계] 다이스 미션 제출합니다.

moondev03 opened the pull request:
# 셀프 체크리스트

<!--
모든 항목을 점검한 후 리뷰를 요청하면, 보다 의미 있는 피드백을 받을 수 있습니다.
-->

- [x] 프로그램이 정상적으로 작동하는가?
- [x] 모든 테스트가 통과하는가?
- [x] 이전에 받은 피드백을 모두 반영하였는가?
- [x] 코딩 스타일 가이드를 준수하였는가?
    - [x] IDE 코드 자동 정렬을 적용하였는가?
    - [x] 린트 검사를 통과하였는가?
- [x] 프로그래밍 요구 사항을 준수하였는가?
- [x] `README.md`에 기능 목록을 정리하고 명확히 기술하였는가?

# 어떤 부분에 집중하여 리뷰해야 할까요?

<!--
학습 목표에 맞게 구현하였는지 점검하고, 피드백을 받을 준비가 되었는지 확인하세요.
정답을 묻기보다, 자신의 생각과 의사 결정 과정을 공유하며 피드백을 통해 논의하는 것이 좋습니다.
-->

### 1단계 피드백 반영
> 수동 로또 번호 생성기가 생긴다면 수동 로또 번호 생성기도 1~45까지의 숫자를 가지고 있어야 할 것처럼 보여요.
클래스를 객체의 능동적인 관리자 역할을 수행할 수 있도록 바꿔보는 것은 어떨까요?
필요에 따라 객체를 꺼내쓸 수 있도록 말이죠. 🤠

538065ca75061ca09dd7350f15618eaf838c9e75

<br>

> 재사용할 수 있도록 만들어보면 어떨까요?

a34eb5aa12bf8afe82157c957d02b887e533d90c,  d03de8b39a1248422e648c98dd6c25de3f2aadc1

<br>

안녕하세요, 크롱 👋
2단계 미션도 잘 부탁드립니다!

moondev03 committed:
chore: 테스트를 위한 고정 번호 생성기 위치 변경

moondev03 committed:
chore: 함수 매개변수 네이밍 수정

moondev03 committed:
refactor: Lotto의 isMatchBonusNumber 함수를 재사용하도록 변경

moondev03 committed:
refactor: LottoNumber 인스턴스 재사용 방식 변경

moondev03 committed:
refactor: WinningNumbers 검증 시 기존 함수 활용하도록 변경

moondev03 committed:
refactor: 총 수익률 테스트 변경

- MethodSource -> CsvSource
- ParameterizedTest 활용해 테스트명 설정

moondev03 committed:
refactor: 테스트 코드 형식 변경

- give, when, then
- FixedLottoNumbersGenerator 사용 제거

moondev03 committed:
refactor: 로또 번호 정렬 테스트 코드 변경

- FixedLottoNumberGenerator 사용 대신 isSorted() 사용

moondev03 committed:
docs(README): 2단계 요구 사항 작성

moondev03 committed:
feat: 로또 번들 결합 기능 구현

moondev03 committed:
feat: 수동 로또 구매 기능 구현

- 수동으로 구매할 개수 만큼 직접 번호를 입력하는 기능
- 직접 입력받은 번호로 로또를 생성하는 기능

moondev03 committed:
refactor: Companion Object 접근 제어자 수정

- Private -> Public
- const val -> private const val

moondev03 committed:
chore: Input Message 상수 수정

- 개행 처리

moondev03 committed:
refactor: PurchaseAmount 활용해서 구매 개수 검증 처리하도록 수정

- LottoMachine 생성 count가 0이면 null 반환하도록 수정
- PurchaseAmount 금액 차감 기능 구현

moondev03 committed:
fix: 구입 금액 0일 때 처리를 위해 combine 단계 위치 조정

moondev03 committed:
feat: 예외처리를 통해 재입력 기능 구현

- 예외 메시지 출력 기능
- 재입력 기능
- 재입력 기능 추가를 위한 Controller 함수 분리 작업

moondev03 committed:
chore: 수동 로또 생성 시 검증 if throw 대신 require로 변경

krrong commented on src/test/kotlin/lotto/domain/model/LottoTest.kt:
```diff
@@ -48,21 +48,21 @@ class LottoTest {
 
         val expected = false
-        val actual = lotto.isMatchBonusNumber(bonusNumber)
+        val actual = lotto.isMatchNumber(bonusNumber)
```
테스트명과 변수명도 무엇을 테스트하는지에 맞춰 변경하면 좋겠어요 🤠

krrong commented on src/test/kotlin/lotto/domain/model/LottoBundleTest.kt:
```diff
@@ -18,4 +21,22 @@ class LottoBundleTest {
+        "3, 11, 14",
+    )
+    @ParameterizedTest(name = "크기가 {0}인 로또 번들과 크기가 {1}인 로또 번들을 결합하면 크기가 {2}인 로또 번들이 되야한다 ")
+    fun `로또 번들을 결합할 수 있다`(
```
ParameterizedTest를 사용하지 않고 한 번만 검증해도 충분해보여요!

krrong commented on src/main/kotlin/lotto/domain/service/RandomLottoNumbersGenerator.kt:
```diff
@@ -3,11 +3,9 @@ package lotto.domain.service
+        return selectedNumbers.map { LottoNumber.from(it) }.toSet()
     }
 
     private fun selectLottoNumbers(): List<Int> {
```
로또의 최대, 최소 값이 LottoNumber, RandomLottoNumbersGenerator 모두 가지고 있어요.
최대 최소 값은 누구의 관심사일까요? 🤔

LottoNumber에서 만들어놓은 값을 잘 활용하려면 어떻게 해야 할까요?

moondev03 replied:
`LottoNumber` companion object에 정의된 `최소 최대 번호 값`을 `public`으로 변경하여
`RandomLottoNumbersGenerator`에서 사용하도록 수정하는게 좋을 것 같습니다.

krrong commented on src/main/kotlin/lotto/controller/LottoController.kt:
```diff
@@ -14,30 +15,96 @@ class LottoController(
-        val purchaseLottoCount = purchasePrice.calculatePurchaseLottoCount(LottoMachine.LOTTO_PRICE)
-        val lottoBundle = LottoMachine().generateLottoBundle(purchaseLottoCount)
+    private fun purchaseLotto(): Pair<LottoBundle?, LottoBundle?> {
+        val purchaseAmount = readPurchaseAmount()
```
`Pair`로 반환을 하고 있는데 게다가 값이 있을지도 없을지도 모른다..?
로또를 구매하는 일을 한다고 생각해보고 조금 더 명확한 값을 반환해보도록 할까요? 🤠

krrong commented on src/main/kotlin/lotto/controller/LottoController.kt:
```diff
@@ -14,30 +15,96 @@ class LottoController(
+        val manualLottoBundle = purchaseManualLotto(purchaseAmount)
+        val autoLottoBundle = purchaseAutoLotto(purchaseAmount)
+
+        return Pair(manualLottoBundle, autoLottoBundle)
```
nit: Pair를 만들 떄는 `a to b` 처럼 사용할 수도 있어요!

krrong commented on src/main/kotlin/lotto/domain/model/LottoMachine.kt:
```diff
@@ -6,8 +6,10 @@ import lotto.domain.service.RandomLottoNumbersGenerator
 class LottoMachine(
     private val generator: LottoNumbersGenerator = RandomLottoNumbersGenerator(),
 ) {
```
디폴트 파라미터를 정의해두면 어떤 문제가 생길 수 있을까요?

moondev03 replied:
디폴트 파라미터로 `RandomLottoNumbersGenerator`를 지정하면, 자동 로또만 사용할 때는 문제가 없지만, 
수동 로또 생성기와 자동 로또 생성기를 모두 사용해야 하는 상황에서는 실수할 가능성이 있을 것 같습니다.

생성자를 통해 명시적으로 `generator`를 전달하도록 하면, 의도한 생성 방식을 확실하게 지정할 수 있어 더 명확한 코드가 될 것 같습니다.

krrong commented on src/main/kotlin/lotto/domain/service/ManualLottoNumbersGenerator.kt:
```diff
@@ -0,0 +1,17 @@
+package lotto.domain.service
```
service 패키지는 어떤 역할을 하나요? 
model과는 어떤 차이가 있기에 패키지를 구분했나요? 🤔

moondev03 replied:
현재 `service` 패키지에 위치한 `generator`들은 데이터를 저장하거나 상태를 관리하지 않으므로 `model`과는 구분된다고 생각합니다.
`generator`의 핵심 역할이 특정 방식에 따라 번호를 생성하는 것이므로, `strategy` 또는 `generator` 패키지로 변경하는 것이 더 적절하다고 생각해 수정하도록 하겠습니다!

krrong commented on src/main/kotlin/lotto/domain/model/PurchaseAmount.kt:
```diff
@@ -9,8 +12,14 @@ class PurchaseAmount(private val amount: Int) {
         return amount / lottoPrice
     }
 
+    fun reducePurchaseAmount(reduceAmount: Int) {
+        require(this.amount >= reduceAmount) { INVALID_REDUCE_AMOUNT_MESSAGE }
+        this.amount -= reduceAmount
+    }
```
자동로또와 수동로또의 구매 수를 계산하기 위해 금액을 줄이는 전략을 사용했네요!

다음과 같은 클래스를 만드는 전략은 어떤가요? 🤔
1. 구매 수량 클래스
  - 수동 구매 수
  - 자동 구매 수
2. 주문 클래스
  - 금액
  - 수동 구매 수
  - 자동 구매 수

krrong commented on src/test/kotlin/lotto/domain/model/LottoResultTest.kt:
```diff
@@ -2,34 +2,24 @@ package lotto.domain.model
 class LottoResultTest {
-    @ParameterizedTest
-    @MethodSource("calculateTotalReturnTest")
+    @ParameterizedTest(name = "로또 등수가 [{0}]일 때, 수익률은 \"{1}\"이다")
+    @CsvSource(
+        "FIRST, 2000000.00",
+        "SECOND, 30000.00",
+        "THIRD, 1500.00",
+        "FOURTH FIFTH, 27.50",
+        "MISS, 0.00",
+        "THIRD FIFTH SECOND, 10501.66",
+    )
     fun `총 수익률을 계산할 수 있다`(
```
테스트를 보다 `split`이 왜 필요할까 하는 의문점이 들었어요.
수익률을 계산하는 테스트를 진행하기 위해서는 테스트의 가독성을 위해 
`CsvSource`를 사용하는 것보다 분리해서 작성하는게 좋아보여요.

krrong reviewed (changes requested):
안녕하세요 다이스
피드백 반영, 2단계 미션 고생하셨어요 👏

질문에 대한 답을 스스로 찾고 계신 것 같아 보기 좋습니다.
고민해보면 좋을 부분들을 코멘트로 달아드렸으니
확인해주시고, 어려운 점이 있다면 언제든 연락주세요 🤠

moondev03 committed:
chore: 패키지명 변경 (service -> generator)

moondev03 committed:
refactor: LottoMachine Default Parameter 제거

moondev03 committed:
refactor: 로또 최소/최대 번호 값 LottoNumber에서 관리하도록 수정

- LottoNumber의 최소/최대 값 public으로 변경
- RandomLottoNumbersGenerator의 최소/최대 값 제거

moondev03 committed:
refactor: 총 수익률 계산 테스트 가독성 개선

moondev03 committed:
refactor: LottoBundle Combine 테스트 케이스 1개로 수정

moondev03 committed:
refactor: 보너스 번호 보유 여부 테스트 가독성 개선

moondev03 committed:
feat: 구입 금액에 따른 수동/자동 구매수를 계산하는 기능

- PurchaseAmount를 줄이는 전략 대신 Order 사용하도록 수정
- Controller 변경

moondev03 commented:
> 자동로또와 수동로또의 구매 수를 계산하기 위해 금액을 줄이는 전략을 사용했네요!
다음과 같은 클래스를 만드는 전략은 어떤가요? 🤔 (구매 수량 클래스 / 주문 클래스)

기존의 금액을 줄이는 전략보다 `주문 클래스`를 통해 주문을 바탕으로 생성하도록 하는게 더 좋은 것 같다는 생각을 해서 수정하였습니다.
이렇게 수정하니 다른 문제인 `Pair<LottoBundle?, LottoBundle?>`도 자연스럽게 해결되었네요 :)

krrong commented on src/test/kotlin/lotto/domain/model/LottoTest.kt:
```diff
@@ -46,24 +46,24 @@ class LottoTest {
 
     @Test
-    fun `보너스 번호가 로또에 존재하면 true를 반환한다`() {
+    fun `1-6까지의 번호를 가진 로또의 번호에 보너스 번호 1이 포함되면 true를 반환한다`() {
```
사실 https://github.com/woowacourse/kotlin-lotto/pull/141#discussion_r1972510071 의 의도는 `로또가 x를 포함하면 true를 반환한다`와 같이 수정하기를 바랐어요.
`Lotto`의 관점에서 바라봤을 때 들어오는 값이 보너스 번호인지 그냥 번호인지 알 필요가 없으니까요 🤠

krrong commented on src/test/kotlin/lotto/domain/model/LottoResultTest.kt:
```diff
@@ -4,10 +4,11 @@ import org.assertj.core.api.Assertions.assertThat
 class LottoResultTest {
-    @ParameterizedTest
-    @MethodSource("calculateTotalReturnTest")
+    @ParameterizedTest(name = "로또 등수가 {0}일 때, 수익률은 \"{1}\"이다")
+    @MethodSource("provideLottoRanksAndReturns")
     fun `총 수익률을 계산할 수 있다`(
```
조금 더 가독성을 개선하고 싶다면 아래처럼 작성할 수도 있겠어요 🤠

```kotlin
@Test
fun `1등의 수익률은 2000000이다`() {
    val result = mapOf(
        LottoRank.FIRST to 1
    )
    val actual = LottoResult(result).calculateTotalReturn(1000).toDouble()
    assertThat(actual).isEqualTo(2000000.00)
}
```

krrong commented on src/main/kotlin/lotto/domain/model/LottoMachine.kt:
```diff
@@ -1,13 +1,14 @@
+import lotto.domain.generator.LottoNumbersGenerator
 
 class LottoMachine(
-    private val generator: LottoNumbersGenerator = RandomLottoNumbersGenerator(),
+    private val generator: LottoNumbersGenerator,
 ) {
```
이 부분은 구조적인 변경이 커질수도 있을 것 같아 참고만 하셔도 좋을 것 같아요 🤠


현재 `LottoMachine`은 `LottoNumbersGenerator`에게 받아온 숫자들을 로또로 만들어주고 있는데요.
현실 세계에 대입해보면 로또는 기계에게 로또 몇 장 줘! 라고 이야기해서 로또를 받아와요.
즉, `LottoMachine`이 Lotto를 혹은 `LottoBundle`을 발급한다! 라고도 볼 수 있죠.
그리고 로또를 파는 상점에서는 돈과 어떤 로또 기계를 사용할지를 받아 로또를 판다.

이렇게 구성하면 `LottoMachine`의 뎁스가 낮아지고 
`ManualLottoNumbersGenerator`는 불필요한 count 멤버변수를 갖지 않을 수 있어요!

krrong commented on src/main/kotlin/lotto/domain/model/Order.kt:
```diff
@@ -0,0 +1,17 @@
+package lotto.domain.model
+
+class Order(
+    purchaseAmount: PurchaseAmount,
+    val manualLottoAmount: Int,
+) {
```
실제로 로또를 사러가서 돈, 수동 번호를 주며 자동 몇개요~ 하고 주문하는 것처럼 
`Order`에 수동 로또 번호도 추가해보면 어떨까요? 🤔

moondev03 replied:
좋은 의견 감사합니다! 😊  

피드백을 반영하기 전에, **Order가 수동 로또 번호까지 포함하는 것이 적절한지** 고민해봤습니다.  
고민해본 결과, 아래와 같은 이유로 Order가 수동/자동 로또 구매 개수만을 관리하는게 좋다고 생각했습니다.

---

### 1. `Order`의 책임이 과도하게 증가한다고 생각합니다.
- 현재 `Order`는 **구매 금액을 기준으로 로또 구매 개수를 계산**하는 역할을 맡고 있습니다.  
- 여기에 **수동 로또 번호까지 포함**하게 되면, `Order`가 단순한 구매 정보를 넘어 로또 번호까지 직접 관리하는 구조가 됩니다.  
- 이는 **객체의 역할이 불필요하게 확장**되는 결과를 초래할 수 있다고 생각합니다.

---

### 2. 변경이 어려운 구조로 발전할 가능성이 높다고 생각합니다.
- 만약 `Order`가 수동 로또 번호까지 포함하도록 변경되면, 향후 로또 구매 로직이 변경될 때 `Order`의 수정이 필수적입니다.  
- 예를 들어, 추후에 반자동 구매 기능이 추가된다면 `Order`가 더 복잡해질 가능성이 높습니다.  
- 현재처럼 **`Order`는 구매 개수와 관련된 계산만 담당**하고,**수동/자동 로또를 별도로 생성하는 방식이 유지보수에 더 유리**하다고 생각합니다.  

---

### 결론
`Order` 객체는 **로또 구매 개수만을 관리**하고, **수동 로또 번호는 별도의 단계에서 처리**하는 것이 더 적절하다고 생각합니다.  
이렇게 하면 **객체 간 역할이 명확해지고, 변경에도 유연하게 대응할 수 있는 구조**를 유지할 수 있다고 생각합니다.

krrong commented on src/main/kotlin/lotto/controller/LottoController.kt:
```diff
@@ -14,30 +17,82 @@ class LottoController(
+        order: Order,
+        lottoBundle: LottoBundle,
+    ) {
+        outputView.printPurchaseLottoCount(order.manualLottoAmount, order.autoLottoAmount)
+        lottoBundle.lottos.forEach { lotto -> outputView.printPurchaseLottoNumbers(lotto.numbers.toList()) }
```
로또 구매 수를 출력할 때는 `Int`를 따로따로 넘겨주고, 로또 번호를 출력할 때는 `List<LottoNumber>`의 형태를 넘겨주고 있네요.
각각의 기준이 있는걸까요? 🤔

moondev03 replied:
고민해보지 않고 작성한 것 같습니다.

`OutputView`에 데이터를 넘길 때, 개인적으로는 `List`보다는 `객체`를 사용하는 것이 더 낫다고 생각합니다.

- 객체는 데이터를 더 직관적으로 표현할 수 있기 떄문에 가독성 측면에서의 이점이 있다고 생각합니다.
- 객체는 나중에 데이터 구조를 변경하기에 더 용이하고, `List`는 그런 점에서 덜 유연하다고 생각합니다.

위와 같은 이유로 `OutputView`에서는 모두 객체를 받도록 수정했습니다.

<br>

다만, 제 의견에 대한 다른 의견은 없는지 크롱의 의견도 궁금합니다!

krrong replied:
좋은 의견인 것 같아요. 
`OutputView`에서도 당연히 `Model`을 받아 사용할 수 있어요.
저 또한 `Model`을 넘겨주는 것이나 `List`를 넘겨주는 것 모두 가능하고,
둘 중 어느 방법을 선택해야 하느냐는 트레이드 오프의 영역이라고 생각해요.

그런데, 다이스의 설계에서 `OutputView`는 `Model`을 알고 있는데, `InputView`는 왜 `Model`을 알지 못할까요?

이 부분은 앞으로 개발을 하면서 고민을 해보시면 좋을 것 같아요 😊

krrong commented on src/test/kotlin/lotto/domain/model/LottoMachineTest.kt:
```diff
@@ -17,19 +20,44 @@ class LottoMachineTest {
     }
 
-    @Test
+    @RepeatedTest(5)
     fun `생성된 로또 번호는 정렬되어야 한다`() {
```
이 테스트는 정말 LottoMachine의 테스트가 맞을까요? 🤔

moondev03 replied:
피드백을 반영하면서 정렬은 `LottoMachine`의 책임이 맞다는 생각을 하게 되었습니다. 
그런데 현재 코드에서는 `각 로또 번호 생성기`에서 정렬을 수행하고 있기 때문에 이를 수정할 필요성을 느꼈습니다.

그래서 아래와 같은 작업을 수행했습니다:

- 로또 번호는 생성 방식에 관계없이 항상 정렬되어야 하므로, 번호 정렬의 책임을 `LottoMachine`으로 이전했습니다.
- Set 변환 후 크기에 대한 검증은 `Lotto`에서 처리하므로 `LottoNumbersGenerator`에서 Set 변환을 제거했습니다.
- `LottoNumbersGenerator`는 이제 `List<LottoNumber>`를 반환하도록 변경했습니다.

krrong replied:
의도했던 바를 정확히 파악해주셨네요 💯

krrong commented on src/main/kotlin/lotto/domain/generator/RandomLottoNumbersGenerator.kt:
```diff
@@ -1,23 +1,19 @@
-        val shuffledNumbers = (MIN_LOTTO_NUMBER..MAX_LOTTO_NUMBER).shuffled()
+        val shuffledNumbers = (LottoNumber.MIN_LOTTO_NUMBER..LottoNumber.MAX_LOTTO_NUMBER).shuffled()
         return shuffledNumbers.take(LOTTO_NUMBER_SIZE)
     }
```
`LottoNumber`의 최대, 최소 값은 `LottoNumber`만이 관심있는 값이죠.
자동 숫자 생성기는 `LottoNumber`에게 이렇게 요청할 수 있겠어요.
`너가 가진 모든 번호를 주겠니?` 그러면 인스턴스화 시켜놓은 값들을 적절히 활용할 수 있을 것 같아요 🤔

krrong reviewed (changes requested):
안녕하세요 다이스!

코드가 많이 나아진게 느껴져요 👏
고민해보면 좋을 부분에 코멘트를 남겨드렸으니
확인 후 리뷰요청 해주세요! 🙇‍♂️

✚ 도전해보면 좋을만한 코멘트도 있는데, 이 부분은 선택사항이니 그런 구현 방향도 있겠구나~ 하고 넘어가도 됩니다 :)

moondev03 committed:
chore: LottoTest isMatchNumber 테스트명 수정

moondev03 committed:
refactor: OutputView 함수들 파라미터 수정

- List -> 객체

moondev03 committed:
refactor: LottoMachine으로 번호 정렬 책임 이동

- 로또 번호는 생성 방식에 관계없이 항상 정렬되어야 하므로, 번호 정렬 책임을 LottoMachine으로 이전
- LottoNumbersGenerator는 List<LottoNumber>를 반환하도록 변경
- Set 변환 이후 크기에 대한 검증은 Lotto에서 처리되므로 LottoNumbersGenerator에서 Set 변환을 제거

moondev03 committed:
refactor: 랜덤 로또 번호 생성 시 LottoNumber의 NUMBERS를 활용하도록 수정

- 번호의 최소 최대 상수 값 private로 변경
- NUMBERS public으로 수정

krrong commented on src/main/kotlin/lotto/domain/generator/RandomLottoNumbersGenerator.kt:
```diff
@@ -0,0 +1,11 @@
+import lotto.domain.model.LottoNumber
+
+class RandomLottoNumbersGenerator : LottoNumbersGenerator {
+    override fun generate(): List<LottoNumber> {
+        val selectedNumbers = LottoNumber.NUMBERS.keys.shuffled().take(Lotto.LOTTO_NUMBER_SIZE)
+        return selectedNumbers.map { LottoNumber.from(it) }
```
모든 LottoNumber를 가져오는 것 또한 메시지를 넘길 수 있겠네요 🤠

krrong commented on src/main/kotlin/lotto/domain/model/LottoMachine.kt:
```diff
@@ -1,13 +1,25 @@
+        return LottoBundle(lottos)
+    }
+
+    private fun generateLottos(count: Int): List<Lotto> {
+        return List(count) { generateLotto() }
+    }
```
`generateLottoBundle` 안에 포함되어도 될 것처럼 보이네요 🤠

krrong commented on src/main/kotlin/lotto/domain/model/PurchaseAmount.kt:
```diff
@@ -1,6 +1,9 @@
 package lotto.domain.model
 
-class PurchaseAmount(private val amount: Int) {
+class PurchaseAmount(amount: Int) {
+    var amount: Int = amount
+        private set
```
amount는 외부에서 사용하지 않으니 가시성을 수정해볼 수 있겠어요

krrong reviewed (approved):
안녕하세요 다이스
로또 미션동안 고생하셨어요! 👏
이번 미션에서 배운 것들을 다음 미션에도 잘 적용했으면 좋겠어요.
앞으로도 응원합니다 💪
